File: src/onomedit/core/collection.py
"""数据管理：文件数组构建、子文件夹展开、排除过滤、重命名排序。"""

from __future__ import annotations

import glob
import os
import sys

from onomedit.core.pathitem import PathItem
from onomedit.utils import clipboard as clipboard_util
from onomedit.utils import fileattr

# 重命名顺序可选项目（默认不排序）
SORT_BY_DEFAULT = "default"
SORT_BY_NAME = "name"
SORT_BY_PATH = "path"
SORT_BY_MTIME = "mtime"
SORT_BY_CTIME = "ctime"
SORT_BY_SIZE = "size"
SORT_BY_CHOICES: tuple[str, ...] = (
    SORT_BY_DEFAULT,
    SORT_BY_NAME,
    SORT_BY_PATH,
    SORT_BY_MTIME,
    SORT_BY_CTIME,
    SORT_BY_SIZE,
)


def read_stream_paths(stream=None) -> list[str]:
    """从 stdin 逐行读取路径（管道输入），清洗空白并跳过空行。

    配合 shell 管道：``dir /b | onomedit rename`` 之类把外部命令输出直接当作
    待重命名路径列表。相对路径按进程当前工作目录解析（与 glob 一致）。
    """
    stream = stream if stream is not None else sys.stdin
    return [line.strip() for line in stream if line.strip()]


def collect_paths(
    raw_paths: list[str] | None, *, use_clipboard: bool = True
) -> list[str]:
    """收集文件路径：参数优先，否则读剪贴板；过滤不存在的路径。

    Windows 上 shell 不展开通配符，程序内主动展开（glob）。
    """
    paths: list[str] = []
    if raw_paths:
        paths.extend(raw_paths)
    elif use_clipboard:
        # 剪贴板：优先 CF_HDROP（资源管理器复制文件/文件夹），否则解析文本路径
        paths.extend(clipboard_util.get_paths())
    expanded: list[str] = []
    for p in paths:
        if glob.has_magic(p):
            expanded.extend(glob.glob(p))
        else:
            expanded.append(p)
    return [p for p in expanded if os.path.exists(p)]


def build_items(paths: list[str]) -> list[PathItem]:
    return [PathItem(p) for p in paths]


def dedupe_items(items: list[PathItem]) -> list[PathItem]:
    """按规范化绝对路径去重，保留首次出现的条目。

    Windows 文件系统大小写不敏感，且 ``a.txt`` 与 ``.\\a.txt``、``./A.TXT`` 均
    指向同一文件，故以 ``normcase(abspath(...))`` 为去重键（POSIX 上 normcase
    原样返回，大小写不同的文件不会被误合并）。
    """
    seen: set[str] = set()
    out: list[PathItem] = []
    for item in items:
        key = os.path.normcase(os.path.abspath(item.full))
        if key in seen:
            continue
        seen.add(key)
        out.append(item)
    return out


def expand_subdirs(items: list[PathItem], depth: int) -> list[PathItem]:
    """把目录项展开为其（限层级）子内容：depth=1 为直接子项，depth=2 再含下一层；
    depth<=0 不展开；展开后的目录项本身不再保留。"""
    if depth <= 0:
        return items
    out: list[PathItem] = []
    for item in items:
        p = item.full
        if not os.path.isdir(p):
            out.append(item)
            continue
        for root, dirs, files in os.walk(p):
            rel = os.path.relpath(root, p)
            level = 0 if rel == "." else rel.count(os.sep) + 1
            if level >= depth:
                # 到达层级边界：剪枝（其内容属于更深层级），本层内容不加入
                dirs[:] = []
                continue
            for d in dirs:
                out.append(PathItem(os.path.join(root, d)))
            for f in files:
                out.append(PathItem(os.path.join(root, f)))
    return out


def display_base(paths: list[str]) -> str:
    """确认窗口显示基准：输入路径的公共父目录。

    单目录输入再向上取一级（选 ``C:\\a\\b`` 展开 → 显示 ``b\\...``）；
    无法计算（不同盘符等）时返回空串。"""
    if not paths:
        return ""
    abs_paths = [os.path.abspath(p) for p in paths]
    try:
        common = os.path.commonpath(abs_paths)
    except ValueError:
        return ""
    if len(paths) == 1:
        common = os.path.dirname(common)
    return common


def sort_items(
    items: list[PathItem], sort_by: str = SORT_BY_DEFAULT, reverse: bool = False
) -> list[PathItem]:
    """按指定项目排序重命名顺序；default 或未知值保持原顺序（不复制列表）。

    - reverse=True 时：default 下反转原顺序，其余按排序键降序；
    - 名称/路径排序用 ``normcase`` 保证 Windows 上大小写不敏感且结果可预测；
      时间/大小排序取 stat 值，stat 失败（权限等）时按 0 处理不中断流程。
    """
    if sort_by == SORT_BY_DEFAULT or sort_by not in SORT_BY_CHOICES:
        if reverse:
            return list(reversed(items))
        return items
    if sort_by == SORT_BY_NAME:
        key = lambda i: os.path.normcase(i.name)
    elif sort_by == SORT_BY_PATH:
        key = lambda i: os.path.normcase(i.full)
    elif sort_by == SORT_BY_MTIME:
        key = lambda i: _stat_attr(i.full, "st_mtime")
    elif sort_by == SORT_BY_CTIME:
        key = lambda i: _stat_attr(i.full, "st_ctime")
    else:  # SORT_BY_SIZE
        key = lambda i: _stat_attr(i.full, "st_size")
    return sorted(items, key=key, reverse=reverse)


def _stat_attr(path: str, attr: str):
    try:
        return getattr(os.stat(path), attr)
    except OSError:
        return 0.0


def apply_excludes(items: list[PathItem], exclude) -> list[PathItem]:
    """应用六类排除开关（文件/目录/符号链接/只读/隐藏/系统）。

    符号链接用 ``islink`` 而非存在性；隐藏/系统/只读 Windows 按属性位，
    其他平台由 fileattr 降级。
    """
    out: list[PathItem] = []
    for item in items:
        p = item.full
        if exclude.files and os.path.isfile(p):
            continue
        if exclude.dirs and os.path.isdir(p):
            continue
        if exclude.symlinks and os.path.islink(p):
            continue
        if exclude.readonly and fileattr.is_readonly(p):
            continue
        if exclude.hidden and fileattr.is_hidden(p):
            continue
        if exclude.system and fileattr.is_system(p):
            continue
        out.append(item)
    return out

File: crates/onomedit-core/src/collection.rs
use std::cmp::Ordering;
use std::collections::HashSet;
use std::fs;
use std::io::{self, BufRead};
use std::path::{Component, Path, PathBuf};

use glob::{MatchOptions, glob_with};

use crate::config::ExcludeOptions;
use crate::path::PathItem;

pub const SORT_BY_CHOICES: [&str; 6] = ["default", "name", "path", "mtime", "ctime", "size"];

pub fn read_stream_paths(reader: impl BufRead) -> io::Result<Vec<String>> {
    reader
        .lines()
        .map(|line| line.map(|value| value.trim().to_owned()))
        .filter(|line| line.as_ref().is_ok_and(|value| !value.is_empty()))
        .collect()
}

pub fn collect_paths(raw_paths: &[String]) -> Vec<PathBuf> {
    let options = MatchOptions {
        case_sensitive: !cfg!(windows),
        require_literal_separator: false,
        require_literal_leading_dot: true,
    };
    let mut output = Vec::new();
    for raw in raw_paths {
        if has_magic(raw) {
            if let Ok(matches) = glob_with(raw, options) {
                output.extend(matches.filter_map(Result::ok).filter(|path| path.exists()));
            }
        } else {
            let path = PathBuf::from(raw);
            if path.exists() {
                output.push(path);
            }
        }
    }
    output
}

fn has_magic(path: &str) -> bool {
    path.contains(['*', '?', '['])
}

pub fn expand_subdirs(items: Vec<PathItem>, depth: i32) -> Vec<PathItem> {
    if depth <= 0 {
        return items;
    }
    let mut output = Vec::new();
    for item in items {
        if item.full().is_dir() {
            expand_directory(item.full(), 1, depth, &mut output);
        } else {
            output.push(item);
        }
    }
    output
}

fn expand_directory(directory: &Path, level: i32, depth: i32, output: &mut Vec<PathItem>) {
    if level > depth {
        return;
    }
    let Ok(entries) = fs::read_dir(directory) else {
        return;
    };
    let mut directories = Vec::new();
    let mut files = Vec::new();
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() && !path.is_symlink() {
            directories.push(path);
        } else {
            files.push(path);
        }
    }
    output.extend(directories.iter().cloned().map(PathItem::new));
    output.extend(files.into_iter().map(PathItem::new));
    for child in directories {
        expand_directory(&child, level + 1, depth, output);
    }
}

pub fn dedupe_items(items: Vec<PathItem>) -> Vec<PathItem> {
    let mut seen = HashSet::new();
    items
        .into_iter()
        .filter(|item| seen.insert(path_key(item.full())))
        .collect()
}

pub fn path_key(path: &Path) -> String {
    let absolute = if path.is_absolute() {
        path.to_owned()
    } else {
        std::env::current_dir().unwrap_or_default().join(path)
    };
    let mut normalized = PathBuf::new();
    for component in absolute.components() {
        match component {
            Component::CurDir => {}
            Component::ParentDir => {
                normalized.pop();
            }
            other => normalized.push(other.as_os_str()),
        }
    }
    let key = normalized.to_string_lossy().into_owned();
    if cfg!(windows) {
        key.to_lowercase()
    } else {
        key
    }
}

pub fn apply_excludes(items: Vec<PathItem>, exclude: &ExcludeOptions) -> Vec<PathItem> {
    items
        .into_iter()
        .filter(|item| {
            let path = item.full();
            !(exclude.files && path.is_file()
                || exclude.dirs && path.is_dir()
                || exclude.symlinks && path.is_symlink()
                || exclude.readonly && is_readonly(path)
                || exclude.hidden && is_hidden(path)
                || exclude.system && is_system(path))
        })
        .collect()
}

pub fn sort_items(mut items: Vec<PathItem>, sort_by: &str, reverse: bool) -> Vec<PathItem> {
    if sort_by == "default" || !SORT_BY_CHOICES.contains(&sort_by) {
        if reverse {
            items.reverse();
        }
        return items;
    }
    items.sort_by(|left, right| compare(left, right, sort_by));
    if reverse {
        items.reverse();
    }
    items
}

fn compare(left: &PathItem, right: &PathItem, sort_by: &str) -> Ordering {
    match sort_by {
        "name" => string_key(&left.name()).cmp(&string_key(&right.name())),
        "path" => path_key(left.full()).cmp(&path_key(right.full())),
        "mtime" => metadata_time(left.full(), true).cmp(&metadata_time(right.full(), true)),
        "ctime" => metadata_time(left.full(), false).cmp(&metadata_time(right.full(), false)),
        "size" => metadata_size(left.full()).cmp(&metadata_size(right.full())),
        _ => Ordering::Equal,
    }
}

fn string_key(value: &str) -> String {
    if cfg!(windows) {
        value.to_lowercase()
    } else {
        value.into()
    }
}

fn metadata_time(path: &Path, modified: bool) -> std::time::SystemTime {
    fs::metadata(path)
        .ok()
        .and_then(|metadata| {
            if modified {
                metadata.modified().ok()
            } else {
                metadata.created().ok()
            }
        })
        .unwrap_or(std::time::UNIX_EPOCH)
}

fn metadata_size(path: &Path) -> u64 {
    fs::metadata(path)
        .map(|metadata| metadata.len())
        .unwrap_or(0)
}

pub fn display_base(paths: &[String]) -> PathBuf {
    if paths.is_empty() {
        return PathBuf::new();
    }
    let absolute: Vec<PathBuf> = paths
        .iter()
        .map(|path| {
            let path = PathBuf::from(path);
            if path.is_absolute() {
                path
            } else {
                std::env::current_dir().unwrap_or_default().join(path)
            }
        })
        .collect();
    let mut common: Vec<_> = absolute[0].components().collect();
    for path in &absolute[1..] {
        let components: Vec<_> = path.components().collect();
        let length = common
            .iter()
            .zip(&components)
            .take_while(|(a, b)| a == b)
            .count();
        common.truncate(length);
    }
    let mut base: PathBuf = common
        .iter()
        .map(|component| component.as_os_str())
        .collect();
    if paths.len() == 1 {
        base.pop();
    }
    base
}

#[cfg(windows)]
fn attributes(path: &Path) -> u32 {
    use std::os::windows::fs::MetadataExt;
    fs::metadata(path)
        .map(|metadata| metadata.file_attributes())
        .unwrap_or(0)
}

fn is_readonly(path: &Path) -> bool {
    fs::metadata(path).is_ok_and(|metadata| metadata.permissions().readonly())
}

fn is_hidden(path: &Path) -> bool {
    #[cfg(windows)]
    return attributes(path) & 0x2 != 0;
    #[cfg(not(windows))]
    return path
        .file_name()
        .is_some_and(|name| name.to_string_lossy().starts_with('.'));
}

fn is_system(path: &Path) -> bool {
    #[cfg(windows)]
    return attributes(path) & 0x4 != 0;
    #[cfg(not(windows))]
    return false;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expands_with_depth_and_deduplicates() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(dir.path().join("a.txt"), "a").unwrap();
        fs::create_dir(dir.path().join("sub")).unwrap();
        fs::write(dir.path().join("sub").join("b.txt"), "b").unwrap();
        let depth_one = expand_subdirs(vec![PathItem::new(dir.path())], 1);
        assert!(depth_one.iter().any(|item| item.name() == "a.txt"));
        assert!(depth_one.iter().any(|item| item.name() == "sub"));
        assert!(!depth_one.iter().any(|item| item.name() == "b.txt"));
        let duplicate = PathItem::new(dir.path().join("a.txt"));
        assert_eq!(dedupe_items(vec![duplicate.clone(), duplicate]).len(), 1);
    }
}
